src/components/MultiHopTrade/hooks/useTradeExecution.tsx

```diff
@@ -0,0 +1,117 @@
+import { supportsETH, supportsEthSwitchChain } from '@shapeshiftoss/hdwallet-core'
+import { useCallback, useState } from 'react'
+import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingleton'
+import { useErrorHandler } from 'hooks/useErrorToast/useErrorToast'
+import { usePoll } from 'hooks/usePoll/usePoll'
+import { useWallet } from 'hooks/useWallet/useWallet'
+import type { Swapper2, TradeQuote2 } from 'lib/swapper/api'
+import { SwapperName } from 'lib/swapper/api'
+import { lifi } from 'lib/swapper/swappers/LifiSwapper/LifiSwapper2'
+import { thorchain } from 'lib/swapper/swappers/ThorchainSwapper/ThorchainSwapper2'
+import { assertUnreachable, isEvmChainAdapter } from 'lib/utils'
+import { selectPortfolioAccountMetadataByAccountId } from 'state/slices/selectors'
+import { useAppSelector } from 'state/store'
+
+import { TRADE_POLL_INTERVAL_MILLISECONDS } from './constants'
+import { useAccountIds } from './useAccountIds'
+
+export const useTradeExecution = ({
+  swapperName,
+  tradeQuote,
+}: {
+  swapperName: SwapperName
+  tradeQuote: TradeQuote2
+}) => {
+  const [txId, setTxId] = useState<string | undefined>()
+  const [message, setMessage] = useState<string | undefined>()
+  const { poll } = usePoll()
+  const { showErrorToast } = useErrorHandler()
+  const wallet = useWallet().state.wallet
+
+  const { sellAssetAccountId } = useAccountIds({
+    sellAsset: tradeQuote.steps[0].sellAsset,
+  })
+
+  const sellAccountMetadata = useAppSelector(state =>
+    selectPortfolioAccountMetadataByAccountId(state, { accountId: sellAssetAccountId }),
+  )
+
+  const executeTrade = useCallback(async () => {
+    if (!wallet) throw Error('missing wallet')
+    if (!sellAccountMetadata) throw Error('missing sellAccountMetadata')
+
+    const {
+      bip44Params: { accountNumber },
+    } = sellAccountMetadata
+
+    const swapper: Swapper2 = (() => {
+      switch (swapperName) {
+        case SwapperName.LIFI:
+          return lifi
+        case SwapperName.Thorchain:
+          return thorchain
+        case SwapperName.Zrx:
+        case SwapperName.CowSwap:
+        case SwapperName.Osmosis:
+        case SwapperName.OneInch:
+        case SwapperName.Test:
+          throw Error('not implemented')
+        default:
+          assertUnreachable(swapperName)
+      }
+    })()
+
+    try {
+      const stepIndex = 0 // TODO: multi-hop trades require this to be dynamic
+      const chainId = tradeQuote.steps[stepIndex].sellAsset.chainId
+
+      const sellAssetChainAdapter = getChainAdapterManager().get(chainId)
+
+      if (!sellAssetChainAdapter)
+        throw Error(`missing sellAssetChainAdapter for chainId ${chainId}`)
+
+      if (isEvmChainAdapter(sellAssetChainAdapter)) {
+        if (!supportsEthSwitchChain(wallet))
```

Comment:
@woodenfurniture this will break for native, which doesn't support ethSwitchChain, is this intended?

Reply:
u sure? `assertSwitchChain` throws if the wallet doesnt support switching:

```
  async assertSwitchChain(wallet: ETHWallet) {
    if (!wallet.ethGetChainId) return

    const walletChainReference = await wallet.ethGetChainId()
    const adapterChainReference = Number(fromChainId(this.chainId).chainReference)

    // switch chain not needed if wallet and adapter chains match
    if (walletChainReference === adapterChainReference) return

    // error if wallet and adapter chains don't match, but switch chain isn't supported by the wallet
    if (!wallet.ethSwitchChain) {
      throw new Error(
        `wallet does not support switching chains: wallet network (${walletChainReference}) and adapter network (${adapterChainReference}) do not match.`,
      )
    }
```

although my check above is probably redundant in that case. Elsewhere throughout the codebase it doesnt look like any calls of `supportsEthSwitchChain` are used to check before calling `assertSwitchChain` either

Reply:
`assertSwitchChain` and `supportsEthSwitchChain` are two different things, the former superseeding the latter with more checks:

- `supportsEthSwitchChain()` is used to detect chain-switching capabilities in wallets like MetaMask and TallyHo, which support the `wallet_switchEthereumChain` JSON-RPC method to switch their active chain. Note, native doesn't have the notion of an active chain, thus doesn't support `ethSwitchChain()`
- `assertSwitchChain()` is used to assert switch chain capabilities using similar logic as `supportsEthSwitchChain()` *only if applicable* (i.e if the wallet supports the notion of an active chain, and the currently active chain isn't the wallet one) . Note the first line with the early return - since native doesn't have a notion of an active ChainId, it doesn't have an `ethGetChainId` method thus will early bail